src/sinks/util/sink.rs

```diff
@@ -0,0 +1,1193 @@
+        match self.inner.start_send(item)? {
+            AsyncSink::Ready => {
+                self.pending += 1;
+                trace!(message = "submit successful.", pending_acks = self.pending);
+
+                if self.pending >= STREAM_SINK_MAX {
+                    self.poll_complete()?;
+                }
+
+                Ok(AsyncSink::Ready)
+            }
+
+            AsyncSink::NotReady(item) => {
+                trace!("Inner sink applying back pressure.");
+                Ok(AsyncSink::NotReady(item))
+            }
+        }
+    }
+
+    fn poll_complete(&mut self) -> Poll<(), Self::SinkError> {
+        try_ready!(self.inner.poll_complete());
+
+        trace!(message = "Acking events.", acking_num = self.pending);
+        self.acker.ack(self.pending);
+        self.pending = 0;
+
+        Ok(().into())
+    }
+}
+
+// === BatchSink ===
+
+/// A `Sink` interface that wraps a `Service` and a
+/// `Batch`.
+///
+/// Provided a batching scheme, a service and batch settings
+/// this type will handle buffering events via the batching scheme
+/// and dispatching requests via the service based on either the size
+/// of the batch or a batch linger timeout.
+///
+/// # Acking
+///
+/// Service based acking will only ack events when all prior request
+/// batches have been acked. This means if sequential requests r1, r2,
+/// and r3 are dispatched and r2 and r3 complete, all events contained
+/// in all requests will not be acked until r1 has completed.
+pub struct BatchSink<S, B, Request, E = DefaultExecutor> {
+    service: ServiceSink<S, Request>,
+    batch: B,
+    settings: BatchSettings,
+    linger: Option<Delay>,
+    closing: bool,
+    exec: E,
+    _pd: PhantomData<Request>,
+}
+
+impl<S, B, Request> BatchSink<S, B, Request, DefaultExecutor>
+where
+    S: Service<Request>,
+    S::Future: Send + 'static,
+    S::Error: Into<crate::Error> + Send + 'static,
+    S::Response: fmt::Debug,
+    B: Batch<Output = Request>,
+{
+    pub fn new(service: S, batch: B, settings: BatchSettings, acker: Acker) -> Self {
+        BatchSink::with_executor(service, batch, settings, acker, DefaultExecutor::current())
+    }
+}
+
+impl<S, B, Request, E> BatchSink<S, B, Request, E>
+where
+    S: Service<Request>,
+    S::Future: Send + 'static,
+    S::Error: Into<crate::Error> + Send + 'static,
+    S::Response: fmt::Debug,
+    B: Batch<Output = Request>,
+    E: Executor,
+{
+    pub fn with_executor(
+        service: S,
+        batch: B,
+        settings: BatchSettings,
+        acker: Acker,
+        exec: E,
+    ) -> Self {
+        let service = ServiceSink::new(service, acker);
+
+        Self {
+            service,
+            batch,
+            settings,
+            linger: None,
+            closing: false,
+            exec,
+            _pd: PhantomData,
+        }
+    }
+
+    fn should_send(&mut self) -> bool {
+        self.closing || self.batch.len() >= self.settings.size || self.linger_elapsed()
```

Comment:
This seems like the only place we're checking `closing`. Is that sufficient to shut down correctly?

Reply:
Yes, it should ensure we hit the batch is empty branch and thus send out all batches then polling in the inner service to complete.